text/0008-ingest-gem-content.md

```diff
@@ -0,0 +1,332 @@
+- Feature Name: Ingest Gem File Contents
+- Start Date: 2023-01-18
+- RFC PR: (leave this empty)
+- Bundler Issue: (leave this empty)
+
+# Summary
+
+Create a process for the purpose of processing and saving the contents of a gem.
+The process will run each time a gem is uploaded to save information about each file in the gem.
+The saved output will include file size, content type, SHA256 checksum, and relative path within the gem.
+In the interest of simplifying this RFC and the feature work that follows, this RFC focuses on only the ingestion and storage of gem content.
+
+# Motivation
+
+It is our hope that we can provide access to individual file data from gem files across versions in a compact and efficient way that supports future features depending on gem content.
+
+Creating a gem content ingestion and storage system will enable rubygems.org to support the security and useability gems in the Ruby Gems ecosystem.
+The gem content repository will be the authoritative source for the content of gems, accessible as individual files without without requiring downloading and unpacking a gem.
+
+In the near future, the indexed gem content will allow an API for accessing the gem content.
+The API will support both rubygems.org and external projects that wish to use individual gem content.
+Rubygems.org plans to add tools for browsing gem content on rubygems.org or through the `gem` command line interface.
+
+# Guide-level explanation
+
+For the purpose of illustration, I will use the gem `rake` as an example.
+
+When a gem is uploaded, rubygems.org saves it as a tarball file named `rake-13.0.6.gem`.
+A gem may be installed with `gem install` or the `.gem` file can be downloaded using `gem fetch`.
+Downloading the entire gem locally is currently the only way to access the contents of gem.
+
+With the gem content ingestion system in place, gem contents will be saved to S3 and indexed in the rubygems.org database.
+Binary files will be recorded in metadata, but not stored.
+Gem contents will be stored at an S3 key named for the SHA256 of the file to de-duplicate files within a gem.
+Yanking will remove file contents unique to the gem, as expected, without altering unyanked gem contents.
+
+### Background process
+
+In order to ingest the contents of a gem, a background process can be enqueued to index and store the gem contents.
+
+At gem upload time, or on demand, a background worker will download, unpack, record file metadata, index files in the database as rows, and upload each file to a content-only S3 bucket.
+The process will result in a row in the database for each file in a gem, with the relative path, checksum, and size of every file.
```

Comment:
instead of a raw checksum, we should store an SRI string here, to support seamlessly migrating checksum algorithms in the future

Reply:
I can see why that would be nice.

Let me just play devils advocate for a moment (though I don't object to SRI in theory). Migrating would never be all at once. We'd have to re-digest all files to continue to benefit from the deduplication, and that would bust all caches for any gem. The algorithm itself isn't really for the purpose of verifying the file content (though that's an added benefit). It's real purpose is a deterministic algorithm that quickly helps us find duplicates. Maybe in 10 years I'll eat my words when there's a painful algorithm migration in rubygems.org.

Since this a naming convention rather than a critical integrity check I'm not sure we'll need to migrate. If we want to add integrity checks to the manifest later, we can use an additional attribute on each file and continue with our naming.

The only concern that would make me not use SHA256 right now is if there was a credible likelihood of multiple files having the same checksum but different fully valid content within the same gem (or multiple gems if we use global hashes).

Reply:
SRI, at least browser implementations, would seem to encourage sha512 and the format requires base64 encoding. Base64 isn't directly url safe, so using it for urls may not work as smoothly as hex encoding. I like the idea of adding this integrity attribute for each file, but I'm not sure the browser supported format would work for our URLs. We could deviate from the browser format or use a second sha256 that just stands in for the key.

I'm not worried about integrity as much for deduplicating file keys. It's infinitesimally unlikely to have 2 valid textual files in a gem that share a sha256.

Reply:
hex-encoded sha256 seems fine as a starting point. the chance that we will eventually want to migrate to a different hash function for security reasons are basically 100%, so I think we should build it in a way that allows us to migrate hash versions in the future. maybe that just means (in the data model) a column for algorithm and it's an enum with only one value so far.

Reply:
I'd like to settle on some algorithm for naming files, and then add an additional column for integrity hashes later if we deem the current algorithm to be insufficient. It's one thing to add a new column and another thing entirely to rewrite the filename of every file already stored, or otherwise give up deduplication at the time when we switch.